### ISSUE TITLE
Create independent modules

### ISSUE THREAD

**Issue by bodograumann** [gertqin/vuex-class-modules#1]
I really like the way this feature looks. Unfortunately I can not use it as-is.

Afaics with `vuex-class-modules` you always have to import the applications main vuex store object. When the application is composed from multiple packages and the store assembled with `store.registerModule`, this is not possible.

Do you think there is a way to define the vuex module independently of the main store and register it later?

**Comment by gertqin** [comment]
Hi, thanks for showing interest!

So if I understand you correctly you would like to be able to just export the vuex-class-modules, and then register them elsewhere?

It should be quite easy to implement - the only problem is that there in this case will be two different objects you have to deal with:

1.  The object to be registered in the store (the _"real"_ vuex module)
2.  The accessor (proxy) object

and I'm not sure how the API should be here.

One suggestion could be to expose a register function - something like

```ts
// myModule.ts
@Module({ name: "myModule" })
class MyModule {
  // ...
}
export const myModule = new myModule();

// store.ts
import { register } from "vuex-class-module";
import { myModule } from "./myModule.ts";

const store = new Vuex.Store({});
register(store, myModule);
```
What do you think?

**Comment by gertqin** [comment]
Actually simpler and cleaner solution could be to just pass the store in the constructor like this

```ts
// myModule.ts
@Module({ name: "myModule" })
export class MyModule extends VuexModule { // VuexModule has the constructor
  // ...
}

// store.ts
import { MyModule } from "./myModule.ts";

const store = new Vuex.Store({});
export const myModule = new MyModule(store)
```

This way with almost the same API as now, you can decide yourself where you want to register the module. Just bear in mind you will also have to import the module from the place it is registered.

Would that work for you?

**Comment by bodograumann** [comment]
Yes this looks great. Importing the module from the dependent npm package is what I do already.

**Comment by gertqin** [comment]
The only thing you need to be aware of then, is that if you want to access other modules from one module, then you cannot just import the class. You have to import the created instance (which I guess would be quite ugly in your case..). Are your modules referencing other modules/accessing rootState?

**Comment by bodograumann** [comment]
Yes, importing the created module from a different dependent is not possible, because the instantiated module resides in the main application, to which the dependent does not have access.

Currently there is only one such dependence, but it is handled (cleanly I think) by injecting a getter from the vuxe module into the dependent using InversifyJS. This is done in the main application where both dependents are imported.

I haven’t put too much thought into more complex situations yet, but the first idea is to also use InversifyJS to inject whole vuex modules. The next thing which might then be useful would be, to not call `new MyModule(store)`, but to also use InversifyJS to load all the vuex modules and inject the main store.

I don’t have that much experience with inversify, but I think adding `@injectable` and `@inject("store")` to `VuexModule` should not be too difficult. In my experience it is then still possible to use the class regularly by giving the store as an argument to the constructor.

**Comment by gertqin** [comment]
I have released a new version, where the modules are registered through the constructor. I have also added that module references can be instantiated through the constructor, which hopefully should solve the problems around importing other modules.

I have never used InversifyJS, but as far as I can tell, it should not be possible to inject both the store, but also other modules through their API. 

Feel free to try it out, and let me know if anything isn't working as you would expect :)
